### PR TITLE
feat(rust): add auth api

### PR DESCRIPTION
**Pull request by twittner**

### CODE REVIEW

File: implementations/rust/ockam/ockam_node/src/context.rs [hunk 1]
```diff
@@ -52,6 +53,12 @@ pub struct Context {
     access_control: Box<dyn AccessControl>,
 }
 
+impl fmt::Debug for Context {
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_node/src/context.rs]: Is that overlapping with #2591 ?

File: implementations/rust/ockam/ockam_api/src/lib.rs [hunk 2]
```diff
@@ -469,3 +472,105 @@ impl<T: Encode<()>> ResponseBuilder<T> {
...
 }
+
+/// A newtype around `Cow<'_, str>` that borrows from input.
+///
+/// Contrary to `Cow<_, str>` the `Decode` impl for this type will always borrow
+/// from input so using it in types like `Option`, `Vec<_>` etc will not produce
+/// owned element values.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cbor(transparent)]
+pub struct CowStr<'a>(#[b(0)] pub Cow<'a, str>);
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_api/src/lib.rs]: What about putting this to a separate file, or `ockam_core`, or even minicbor? And this is definitely separate commit

File: implementations/rust/ockam/ockam_api/src/auth.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+use types::{Attribute, Attributes};
+
+/// Auth API server.
+#[derive(Debug)]
+pub struct Server<S> {
+    store: S,
+}
+
+#[ockam_core::worker]
+impl<S: Storage + Send + Sync + 'static> Worker for Server<S> {
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_api/src/auth.rs:25]: Pros vs Cons adding `Send + Sync + 'static` as a requirement for the Storage trait itself? Not sure we ever use it without this requirement

**Review by SanjoDeundiak — DISMISSED** [review]
Nice. Few questions
||
v

File: implementations/rust/ockam/ockam_node/src/context.rs (hunk 1 above)
Comment (twittner) [reply implementations/rust/ockam/ockam_node/src/context.rs]: Indeed it is.

File: implementations/rust/ockam/ockam_api/src/lib.rs (hunk 2 above)
Comment (twittner) [reply implementations/rust/ockam/ockam_api/src/lib.rs]: I will move this into a separate PR.

File: implementations/rust/ockam/ockam_api/src/auth.rs (hunk 3 above)
Comment (twittner) [reply implementations/rust/ockam/ockam_api/src/auth.rs:25]: Still more general to constrain it on call site I think.

**Review by mrinalwadhwa — DISMISSED** [review]

**Review by mrinalwadhwa — APPROVED** [review]

**Pull request merged**